### PR TITLE
server: try graceful close connection first

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Ansible uses `terminal` signal to upgrade TiDB.
Currently TiDB will close all connection directly when got `terminal` signal.

This PR make TiDB try to firstly graceful close all connection with a timeout. If timeout, directly closes all connection.

### Check List 

Tests 

Code changes

 - Has exported function/method change

Side effects

 
Related changes

 - Need to cherry-pick to the release branch

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8711)

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -387,21 +388,50 @@ func (s *Server) KillAllConnections() {
 	}
 }
 
+var gracefulCloseConnectionsTimeout = 15 * time.Second
+
+// TryGracefulDown will try graceful close all connection first with timeout. if timeout, will close all connection directly.
```
Comment (tiancaiamao) [inline server/server.go]: Gramma

File: server/server.go [hunk 2]
```diff
@@ -387,21 +388,50 @@ func (s *Server) KillAllConnections() {
...
+func (s *Server) TryGracefulDown() {
+	ctx, cancel := context.WithCancel(context.Background())
+	timeout := time.Tick(gracefulCloseConnectionsTimeout)
```
Comment (tiancaiamao) [inline server/server.go]: I guess you need `time.After` instead of `time.Tick`

Comment (tiancaiamao) [reply server/server.go]: Or `context.WithTimeout` ?

File: server/server.go [hunk 3]
```diff
@@ -387,21 +388,50 @@ func (s *Server) KillAllConnections() {
...
+	ch := make(chan struct{})
+	go func() {
+		s.GracefulDown(ctx)
+		close(ch)
+	}()
+	select {
+	case <-timeout:
+		cancel()
+		s.KillAllConnections()
```
Comment (tiancaiamao) [inline server/server.go]: There is a race condition that `<-timeout` and `s.GracefulDown(ctx)` both triggered.
Will the connection be close twice ?

File: server/server.go [hunk 4]
```diff
@@ -387,21 +388,50 @@ func (s *Server) KillAllConnections() {
...
 
 		count = s.ConnectionCount()
+		if count == 0 {
+			break
+		}
 		// Print information for every 30s.
 		if i%30 == 0 {
 			log.Infof("graceful shutdown...connection count %d\n", count)
 		}
+		time.Sleep(time.Second)
```
Comment (tiancaiamao) [inline server/server.go]: During sleep, the cancel signal can't be catch
How about:
```
ticker := time.Ticker(time.Second)
select {
    case <-ctx.Done():
        return 
    case <- ticker.C:
        break
}
```

**Comment by tiancaiamao** [comment]
Should we adopt this behavior: graceful shutdown first, if it's not closed after 15s, do the normal shutdown process. @shenli @winkyao

**Comment by crazycs520** [comment]
Currently TiDB will directly Close all connection after gets `terminate` signal, The client will get `undeterminate` error, So client has check whether the operation fails or succeed.

This PR can't complement avoid something. Maybe I should close this PR.

**Comment by winkyao** [comment]
@crazycs520 we can avoid some cases, so this PR is valuable.

**Comment by crazycs520** [comment]
OK， let's merge this, @winkyao @jackysp @tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM
Maybe we need a blog or document to explain the `undeterminate` problem and the client retry behavior during TiDB restart @crazycs520

**Pull request merged**